### PR TITLE
Add a specification for importing remote plans

### PR DESCRIPTION
**Pull request by psss**

### CODE REVIEW

File: spec/plans/remote.fmf [hunk 1]
```diff
@@ -0,0 +1,23 @@
+summary: Referencing remote plans
+story:
+    As a tester I want to reference remote plans
+    in order to reuse the config without duplication.
+description:
```
Comment (lukaszachy) [inline spec/plans/remote.fmf:5]: Can remote contain list of dictionaries?
Name has to match exactly one plan, right? Is it regexp (as --name options usually are)?

**Comment by sopos** [comment]
Finally I have a UC for the multiple remotes undef one _symlink plan_:
As we would like to cover Tiers execution by this I would like to have something as follows for e.g. rsyslog:

```yaml
/rsyslog:
    /Sanity:
        /local-messaging:
            description: test local messages delivery
            test: ./runtest.sh
            tag: Tier1
    /Multihost:
        /tcp:
            description: deliver messages via tcp connection
            test: ./runtest.sh
            tag: Tier1multi
    /plans:
        /Tier1single:
            provision:
                how: virtual
            discover:
                how: fmf
                filter: "tag:Tier1"
            execute:
                how: tmt
        /Tier1multi:
            provision:
              - name: client
                how: virtual
              - name: server
                how: virtual
            discover:
                how: fmf
                filter: "tag:Tier1multi"
            execute:
                how: tmt
        /Tier1:
            remote-plan:
              - name: /plans/Tier1single
              - name: /plans/Tier1multi
```
the tier master's plan would contain:
```yaml
/Tier1:
    /rsyslog:
        remote_plan:
            url: https://github.com/RedHat-SP-Security/tests
            name: /rsyslog/plans/Tier1
```

As rsyslog uses both single host and multi host tests for Tier1 testing this could be a way how to _export_ both plans as one entity for Tier masters to handle. They would refer just to one plan per component which is actually the goal here. We do not want to make them to refer to all the component specific stuff (multiple plans base on the component needs - there may various reasons for having more Tier1 plans for a component)
Thus, we need to also support:
1. nesting as we cannot prevent from referencing another _reference_
2. prefixes generation (the final plan name should be made of the _symlink_ name together with the _reference/remote_ name), in case of nested references, the name would hold all of the plan names. e.g. `rsyslog/Tier1/Tier1single` and `rsyslog/Tier1/Tier1multi`.

File: spec/plans/remote.fmf (hunk 1 above)
Comment (psss) [reply spec/plans/remote.fmf]: Only one plan can be used and no matching is applied. The spec updated to be explicit about this.

**Comment by psss** [comment]
@sopos, thank you for providing the detailed example. I understand the use case but I'm afraid your proposal for multiple references under a single plan symlink substantially complicates the implementation. Also, one symlink pointing to multiple locations sounds a bit weird. But the most important point is that it would not be possible to even list names of available plans using `tmt plan ls` without fetching remote stuff which seems to me as an unquestionable requirement.

**Comment by kkaarreell** [comment]
Btw, there is https://pypi.org/project/xunitmerge/ , maybe we could use it to merge individual xunit files from individual runs so that results can be reported together.

**Comment by sopos** [comment]
> @sopos, thank you for providing the detailed example. I understand the use case but I'm afraid your proposal for multiple references under a single plan symlink substantially complicates the implementation. Also, one symlink pointing to multiple locations sounds a bit weird. But the most important point is that it would not be possible to even list names of available plans using `tmt plan ls` without fetching remote stuff which seems to me as an unquestionable requirement.

It may sound weird if we call it a test plan _symlink_. It does not if we call it _a plan of plans_.

If this plan _symlink_ is just another feature, then it's ok. Definitely there are some use cases for it, e.g. system upgrades testing.
For the Tiers testing we would probably want to see _the plan of plans_.

**Comment by psss** [comment]
> Btw, there is https://pypi.org/project/xunitmerge/ , maybe we could use it to merge individual xunit files from individual runs so that results can be reported together.

Thanks for the pointer. This might be useful when handling results locally or for the tier use case. But the important point is that the remote plan support in `tmt` will enable referencing remote plans seamlessly even within CI without any change in the pipelines. Otherwise the results aggregation would have to be implemented/handled there.

**Comment by psss** [comment]
> It may sound weird if we call it a test plan _symlink_. It does not if we call it _a plan of plans_.

Understood, but the question is not about how we call it but how we implement it. And here I see a substantial difference: While plan symlink should be easy (just clone the repo and load the plan from there) the plan of plans brings much more complexity.

> If this plan _symlink_ is just another feature, then it's ok. Definitely there are some use cases for it, e.g. system upgrades testing. For the Tiers testing we would probably want to see _the plan of plans_.

Yes, we can consider this first step as a separate feature. It will allow us to enable integration testing and can be used for tiers as well although with a bit of discomfort:

```yaml
/rsyslog:
    /singlehost:
        plan:
            url: https://github.com/RedHat-SP-Security/tests
            name: /rsyslog/plans/Tier1single
    /multihost:
        plan:
            url: https://github.com/RedHat-SP-Security/tests
            name: /rsyslog/plans/Tier1multi
```

Inheritance could slightly help here as well if needed:

```yaml
/rsyslog:
    plan:
        url: https://github.com/RedHat-SP-Security/tests

    /singlehost:
        plan+:
            name: /rsyslog/plans/Tier1single
    /multihost:
        plan+:
            name: /rsyslog/plans/Tier1multi
```

For the first PoC I'd say it's not that user unfriendly approach.

**Review by lukaszachy — APPROVED** [review]

**Review by happz — APPROVED** [review]

**Review by jscotka — CHANGES_REQUESTED** [review]
definition seems well. Just one idea or maybe question based on example:
```yaml
/plans:
     /plan1:
        execute: 
           how: fmf ...
     /remote_plan:
        plan:
           url: some://url
     /mixed_remote:
       plan:
           url: some://url
       /origin:
       /mix:
           report:
               how: polarion
```

so this example is about, if there will be applied inheritance in this case, and if it is allowed, or in case not, we should explicitly mention this, that it is not allowed.

**Comment by psss** [comment]
> if there will be applied inheritance in this case, and if it is allowed, or in case not, we should explicitly mention this, that it is not allowed.

Good point! At least for the start I'd explicitly state that no other keys besides `plan` are supported. Will update the spec.

File: spec/plans/remote.fmf [hunk 2]
```diff
@@ -0,0 +1,44 @@
+summary: Referencing remote plans
+
+story:
+    As a user I want to reference a plan from a remote repository
+    in order to prevent duplication and minimize maintenance.
+
+description: |
```
Comment (thrix) [inline spec/plans/remote.fmf:7]: @psss should we also describe a bit in detail how will a remote plan inclusion influence the plan where we do the include of the remote plan? Maybe with some examples? Or is that not decided yet?

Comment (psss) [reply spec/plans/remote.fmf:7]: What do you mean by "influence"? The remote plan will be fetched and its steps will replace the `plan` key with the real step configuration. Everything should work completely the same.

**Review by jscotka — APPROVED** [review]
now it seem well, thanks.

File: spec/plans/remote.fmf (hunk 2 above)
Comment (thrix) [reply spec/plans/remote.fmf:7]: ok, I guess I got my answers on the call

**Review by thrix — APPROVED** [review]
LGTM

**Review by ZelenyMartin — APPROVED** [review]

**Review by happz — APPROVED** [review]

**Review by lukaszachy — APPROVED** [review]

**Review by jscotka — APPROVED** [review]
LGTM

**Comment by psss** [comment]
Thanks much to all for patientce and providing valuable feedback!

**Comment by psss** [comment]
Seems github got confused by my repush, the merged commit for reference: https://github.com/teemtee/tmt/commit/0eea512722f85859e01cc52e0410f09c66278aaf

**Pull request merged**